### PR TITLE
correct height of cutouts within a layer (fix #8)

### LINKED ISSUE #8: cutout height
it would be great if the top layer would sit on top of the cutouts.
this would allow me to only use one inset on top of a box with lots of space underneath it.

Example: 5L box with 3 layers, showing top layer:

image

### PR DESCRIPTION
**Pull request by rdmueller**
This small change should end the cutouts for the top layer at the right height.

see #8 for example rendering.

Rendering looks good. Will print a test tomorrow.

### CODE REVIEW

File: IKEA_Samla_Inserts.scad [hunk 1]
```diff
@@ -38,6 +38,11 @@ Layer_Marking_Type = "engrave"; // ["engrave":engrave, "emboss":emboss]
 // Layer marking height
 Layer_Marking_Height = 0.3;
 
+// Cutouts
+// full cutouts are easier to print (no support needed)
+// correct cutouts let the top layer rest on the handle-cutouts
+Cutouts = "full"; //["full": full, "correct", correct]
```
Comment (Alex9779) [inline IKEA_Samla_Inserts.scad]: This is placed in the "Layer marking" section, should be moved up before line 28.
Also this gives me three options in the "Customizer": "full", "correct" and "correct", last comma should be a semi-colon?

**Comment by Alex9779** [comment]
Ok got a minute to have a look.
I am also not really happy with the naming, "correct" seems there was something wrong but this was a decision I made. Initially because for this to work right the cutout height has to be nearly correct which it wasn't and then I implemented the removed cutout if the top layer was smaller. And as told in the issue I didn't think about and consider this a valid special use case.
I would like to see it named like "Top Layer Cutouts", must not be the variable name, and the options "full" and "reduced" maybe (?) or "cut through" and "cutout height limited"... No real idea... :P

Thanks!

**Review by Alex9779 — CHANGES_REQUESTED** [review]

File: IKEA_Samla_Inserts.scad (hunk 1 above)
Comment (rdmueller) [reply IKEA_Samla_Inserts.scad]: ups. typo

**Comment by rdmueller** [comment]
you are right - I was also a bit unsure if "correct" would be a good wording. I will brainstorm a bit about it and come up with a new wording...

File: IKEA_Samla_Inserts.scad (hunk 1 above)
Comment (rdmueller) [reply IKEA_Samla_Inserts.scad]: fixed this

**Comment by rdmueller** [comment]
I now switched to the wording "full" and "partial". I think this refelct the situation quite well.
I also renamed everything to TopLayerCutouts as you suggested.

**Review by Alex9779 — APPROVED** [review]

File: IKEA_Samla_Inserts.scad [hunk 2]
```diff
@@ -25,6 +25,11 @@ Addtional_Spacing = 1;
 // Part (just print part of a layer, halve or quarter)
 Part = "false"; // ["false":false, "halve_column":halve by column, "halve_row":halve by row, "quarter":quarter]
 
+// TopLayerCutouts
+// full TopLayerCutouts are easier to print (no support needed)
+// partial cutouts let the top layer rest on the handle-cutouts
+TopLayerCutouts = "full"; //["full": full, "partial": partial]
```
Comment (Alex9779) [inline IKEA_Samla_Inserts.scad]: The first line should have spaces like the others.
Variable name should be "Top_Layer_Cutouts", underlines are rendered as spaces in the Customizer.
Also the sub-text seems too long, Customizer can't handle two lines:
image
Just the last one is displayed...

Suggestion:
```
// Top Layer Cutouts
// full are easier to print (no support needed), partial cutouts let the top layer rest on the handle-cutouts
Top_Layer_Cutouts = "full"; //["full": full, "partial": partial]
````

**Review by Alex9779 — CHANGES_REQUESTED** [review]
Sorry but it seems this simple approach won't work out.
Switching to partial also shrinks the cutouts a little bit:
Comparison
Hard to see in the animation but it is there.
The problem is the method I use to get the overall shrinkage of the cutouts to the top (also the widening of the box): scaling.
And scaling is calculated over the extrusion length (in this case the height) and a factor (here two factor one for x and one for y).
If you reduce the height and do not modify the factors the shape changes.
I remeber I had that problem initially when trying to just shrink the whole container to get the wall and bottom in one step. I changed then the bottom to be of a different thickness than the walls and so I had to generate it a different way and that problem was gone.
So the cutouts must be extruded the whole height, otherwise you have to adjust the scale factors which I don't think is easy. You could do some math to reduce them.
But I think an approach cutting of the top and regenerate without cutouts would be best.

Suggestion:
```
module Samla_Content(layer, width, depth, height, scale_width, scale_depth, width_handle, depth_handle, width_cutout, depth_cutout, scale_handle, scale_cutout, handle_cutout_height, diameter, diameter2, offset) {
    union() {
        difference() {
            linear_extrude(height=height, scale = [scale_width, scale_depth]) {
                offset(delta=-Addtional_Spacing+offset) Samla_Base(layer, width, depth, height, diameter, width_cutout, scale_cutout);
            }
            Samla_HandleAndCutout(layer, width, depth, height, scale_width, scale_depth, width_handle, depth_handle, width_cutout, depth_cutout, scale_handle, scale_cutout, handle_cutout_height, diameter, diameter2, offset);
        }
        if (Top_Layer_Cutouts == "partial") {
            intersection() {
                translate([0, 0, height-(height-handle_cutout_height)/2+(offset!=0?Bottom_Thickness:0)]) cube([width*scale_width, depth*scale_depth, height-handle_cutout_height], true);
                linear_extrude(height=height, scale = [scale_width, scale_depth]) {
                        offset(delta=-Addtional_Spacing+offset) Samla_Base(layer, width, depth, height, diameter, width_cutout, scale_cutout);
                }
            }
        }
    }
}
```

**Comment by rdmueller** [comment]
I feared something like that. 
I printed 3 of those top layers for 5L boxes and the handle-cutouts did fit quite nice. I only gave the normal cutouts a bit more depth.
What I also noticed with my solution is that (especially for the normal cutouts) there seems to be a slight gap between the vertical and orizontal plane. Nothing to worry about in my case but not perfect.

File: IKEA_Samla_Inserts.scad (hunk 2 above)
Comment (rdmueller) [reply IKEA_Samla_Inserts.scad]: hm. ok. But does the cutomizer work at all? You are speaking of the thingiverse customizer, aren't you? I didn't succeed with it.
Nevertheless, your suggestion is fine. I will fix this.

Comment (Alex9779) [reply IKEA_Samla_Inserts.scad]: I actually meant the Customizer in OpenSCAD...

**Comment by Alex9779** [comment]
> I feared something like that.

Well test my suggestion, no other change needed except the variable definition at the top, with that the partial cutouts don't change compared to full (in any layer).

**Comment by Alex9779** [comment]
Basically I am just doing what I described, I create the content again without cutouts, then a box to just get the upper top, slight hack needed to get the Bottom_Thickness right...

**Comment by Alex9779** [comment]
Still have to revert what you did in Samla_HandleAndCutout....

**Comment by rdmueller** [comment]
ok, I applied your suggestion and got a nice rendering. I also sliced it and it seems that there is no gap between horizontal and vertical plane anymore :-) Thanx for your help!

**Comment by Alex9779** [comment]
> Still have to revert what you did in Samla_HandleAndCutout....

**Comment by rdmueller** [comment]
ups. missed that

**Review by Alex9779 — APPROVED** [review]

**Comment by Alex9779** [comment]
Thanks for your contribution!

**Pull request merged**